### 데이터 불러오기

In [1]:
import numpy as np
import pandas as pd

In [2]:
train = pd.read_csv('pj_data/train.csv', index_col = 'no')
test = pd.read_csv('pj_data/test.csv', index_col = 'no')

train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   
29305   65                  ?   94809        HS-grad              9   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   
29305              Widowed                   ?   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   
29305                White   Female             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  
29305              40   United-States       0  

[29305 rows x 15 columns]

In [3]:
test

age          workclass  fnlwgt      education  education-num  \
no                                                                    
29306   18                  ?  245274   Some-college             10   
29307   29            Private   83003        HS-grad              9   
29308   45            Private   35136      Bachelors             13   
29309   42   Self-emp-not-inc   64631      Bachelors             13   
29310   41            Private  195821      Doctorate             16   
...    ...                ...     ...            ...            ...   
48838   45   Self-emp-not-inc  116789        HS-grad              9   
48839   48            Private  185079        HS-grad              9   
48840   63            Private  117473   Some-college             10   
48841   18            Private  150817           11th              7   
48842   31            Private  341632     Assoc-acdm             12   

            marital-status        occupation    relationship    race      sex  \
no                                                                              
29306        Never-married                 ?       Own-child   White     Male   
29307   Married-civ-spouse     Other-service            Wife   White   Female   
29308   Married-civ-spouse      Tech-support         Husband   Black     Male   
29309   Married-civ-spouse   Exec-managerial         Husband   White     Male   
29310   Married-civ-spouse   Exec-managerial            Wife   White   Female   
...                    ...               ...             ...     ...      ...   
48838   Married-civ-spouse      Craft-repair         Husband   White     Male   
48839        Never-married   Exec-managerial   Not-in-family   White   Female   
48840   Married-civ-spouse    Prof-specialty         Husband   White     Male   
48841        Never-married             Sales       Own-child   White   Female   
48842   Married-civ-spouse      Tech-support         Husband   Black     Male   

       capital-gain  capital-loss  hours-per-week  native-country  
no                                                                 
29306             0             0              16   United-States  
29307             0             0              40   United-States  
29308             0             0              40   United-States  
29309             0             0              40   United-States  
29310             0          1902              40   United-States  
...             ...           ...             ...             ...  
48838             0             0              60   United-States  
48839             0             0              50   United-States  
48840          4386             0              40   United-States  
48841             0             0              20   United-States  
48842             0             0              46   United-States  

[19537 rows x 14 columns]

In [4]:
print(train.shape)
print(test.shape)

(29305, 15)
(19537, 14)


# 데이터 전처리

### 결측치 파악하기

범주형 결측치 파악

In [5]:
train['workclass'].value_counts()                 # workclass 결측치 있음, 4번째로 많은 범주에 해당

 Private             20410
 Self-emp-not-inc     2305
 Local-gov            1868
 ?                    1663
 State-gov            1201
 Self-emp-inc          987
 Federal-gov           854
 Without-pay            12
 Never-worked            5
Name: workclass, dtype: int64

In [6]:
test['workclass'].value_counts()         

 Private             13496
 Self-emp-not-inc     1557
 Local-gov            1268
 ?                    1136
 State-gov             780
 Self-emp-inc          708
 Federal-gov           578
 Without-pay             9
 Never-worked            5
Name: workclass, dtype: int64

In [7]:
train['education'].value_counts()

 HS-grad         9435
 Some-college    6558
 Bachelors       4848
 Masters         1587
 Assoc-voc       1230
 11th            1057
 Assoc-acdm       980
 10th             816
 7th-8th          574
 Prof-school      500
 9th              430
 12th             415
 Doctorate        347
 5th-6th          334
 1st-4th          148
 Preschool         46
Name: education, dtype: int64

In [8]:
train['marital-status'].value_counts()

 Married-civ-spouse       13331
 Never-married             9673
 Divorced                  4030
 Widowed                    942
 Separated                  914
 Married-spouse-absent      394
 Married-AF-spouse           21
Name: marital-status, dtype: int64

In [9]:
train['occupation'].value_counts()                  # occupation  결측치 있음, 중간정도로 많은 범주에 해당

 Prof-specialty       3724
 Craft-repair         3632
 Exec-managerial      3609
 Adm-clerical         3440
 Sales                3292
 Other-service        2975
 Machine-op-inspct    1811
 ?                    1668
 Transport-moving     1393
 Handlers-cleaners    1249
 Farming-fishing       888
 Tech-support          865
 Protective-serv       603
 Priv-house-serv       146
 Armed-Forces           10
Name: occupation, dtype: int64

In [10]:
test['occupation'].value_counts()                  # occupation  결측치 있음, 중간정도로 많은 범주에 해당

 Craft-repair         2480
 Exec-managerial      2477
 Prof-specialty       2448
 Sales                2212
 Adm-clerical         2171
 Other-service        1948
 Machine-op-inspct    1211
 ?                    1141
 Transport-moving      962
 Handlers-cleaners     823
 Farming-fishing       602
 Tech-support          581
 Protective-serv       380
 Priv-house-serv        96
 Armed-Forces            5
Name: occupation, dtype: int64

In [11]:
train['relationship'].value_counts()

 Husband           11769
 Not-in-family      7637
 Own-child          4554
 Unmarried          3071
 Wife               1370
 Other-relative      904
Name: relationship, dtype: int64

In [12]:
train['race'].value_counts()                      

 White                 25081
 Black                  2776
 Asian-Pac-Islander      919
 Amer-Indian-Eskimo      290
 Other                   239
Name: race, dtype: int64

In [13]:
train['sex'].value_counts()

 Male      19621
 Female     9684
Name: sex, dtype: int64

In [14]:
train['native-country'].value_counts()             # native-country 결측치 있음, 3번째로 많은 범주에 해당

 United-States                 26301
 Mexico                          579
 ?                               495
 Philippines                     178
 Germany                         124
 Puerto-Rico                     117
 Canada                          111
 El-Salvador                      96
 Cuba                             94
 India                            90
 England                          75
 South                            74
 Jamaica                          68
 China                            68
 Dominican-Republic               61
 Italy                            60
 Japan                            54
 Columbia                         53
 Poland                           53
 Haiti                            51
 Guatemala                        51
 Vietnam                          50
 Portugal                         43
 Taiwan                           39
 Iran                             38
 Nicaragua                        28
 Greece                           25
 

In [15]:
test['native-country'].value_counts()             # native-country 결측치 있음, 3번째로 많은 범주에 해당

 United-States                 17531
 Mexico                          372
 ?                               362
 Philippines                     117
 Germany                          82
 Canada                           71
 Puerto-Rico                      67
 India                            61
 El-Salvador                      59
 China                            54
 England                          52
 Italy                            45
 Cuba                             44
 Dominican-Republic               42
 South                            41
 Jamaica                          38
 Japan                            38
 Guatemala                        37
 Vietnam                          36
 Poland                           34
 Columbia                         32
 Taiwan                           26
 Haiti                            24
 Portugal                         24
 Greece                           24
 Ecuador                          24
 Peru                             21
 

In [16]:
train['income'].value_counts()

0    22260
1     7045
Name: income, dtype: int64

- 숫자형 데이터 결측치 파악

In [17]:
train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   
29305   65                  ?   94809        HS-grad              9   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   
29305              Widowed                   ?   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   
29305                White   Female             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  
29305              40   United-States       0  

[29305 rows x 15 columns]

In [18]:
is_question = train['age'] == " ?"
train[is_question]                           # 모든 숫자형 컬럼 확인 결과 결측치 없음!!

C:\Users\rnjsx\anaconda3\lib\site-packages\pandas\core\ops\array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income]
Index: []

workclass : 1663, 
occupation : 1668, 
native-country : 495 결측치 존재

### 결측치 채우기 1번 방법 : 결측치 제거

- workclass, occupation, native-country의 결측치 행 제거

In [19]:
# rem_MV(remove missing value)에 workclass가 " ?"인 모든 행을 저장합니다.
# drop함수 이용해서 해당 행 삭제합니다.
rem_MV = train[train['workclass'] == " ?"].index
train = train.drop(rem_MV)

In [20]:
train[train['workclass'] == " ?"]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income]
Index: []

In [21]:
rem_MVt = test[test['workclass'] == " ?"].index
test = test.drop(rem_MVt)

In [22]:
test[test['workclass'] == " ?"]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country]
Index: []

In [23]:
# 같은 방식으로 occupation이 "?"인 모든 행을 삭제합니다
rem_MV1 = train[train['occupation'] == " ?"].index
train = train.drop(rem_MV1)

In [24]:
train[train['occupation'] == " ?"]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income]
Index: []

In [25]:
rem_MVt1 = test[test['occupation'] == " ?"].index
test = test.drop(rem_MVt1)

In [26]:
test[test['occupation'] == " ?"]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country]
Index: []

In [27]:
rem_MV2 = train[train['native-country'] == " ?"].index
train = train.drop(rem_MV2)

In [28]:
train[train['native-country'] == " ?"]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income]
Index: []

In [30]:
rem_MVt2 = test[test['native-country'] == " ?"].index
test = test.drop(rem_MVt2)

In [32]:
test[test['native-country'] == " ?"]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country]
Index: []

### 데이터 컬럼 병합 및 삭제

# 모델 예측

### 문제, 정답 데이터로 분리

In [33]:
# train 데이터에서 정답 컬럼
y_train = train['income']
# train 데이터에서 정답 컬럼을 삭제한 후 남은 모든 컬럼들
X_train = train.drop('income', axis=1)
# 맨 처음에 받았던 test.csv
X_test = test

In [34]:
X_train.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country'],
      dtype='object')

In [35]:
X_test.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country'],
      dtype='object')

In [36]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27173 entries, 1 to 29304
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             27173 non-null  int64 
 1   workclass       27173 non-null  object
 2   fnlwgt          27173 non-null  int64 
 3   education       27173 non-null  object
 4   education-num   27173 non-null  int64 
 5   marital-status  27173 non-null  object
 6   occupation      27173 non-null  object
 7   relationship    27173 non-null  object
 8   race            27173 non-null  object
 9   sex             27173 non-null  object
 10  capital-gain    27173 non-null  int64 
 11  capital-loss    27173 non-null  int64 
 12  hours-per-week  27173 non-null  int64 
 13  native-country  27173 non-null  object
 14  income          27173 non-null  int64 
dtypes: int64(7), object(8)
memory usage: 3.3+ MB


### 데이터 인코딩 하기

In [37]:
X_train = pd.get_dummies(X_train)
X_test = pd.get_dummies(X_test)
# X_train
# X_test

In [38]:
print(X_train.shape)
print(X_test.shape)

(27173, 104)
(18049, 103)


### 컬럼 개수 맞추기

In [39]:
set(X_train.columns) - set(X_test.columns)

{'native-country_ Holand-Netherlands'}

In [40]:
# 'native-country_ Holand-Netherlands'의 컬럼은 생기지만 실제로 컴퓨터가 데이터는 없는 것으로 인식함!
X_test['native-country_ Holand-Netherlands'] = 0

In [41]:
tmp = X_train.columns
for x in tmp :
    print(x)

age
fnlwgt
education-num
capital-gain
capital-loss
hours-per-week
workclass_ Federal-gov
workclass_ Local-gov
workclass_ Private
workclass_ Self-emp-inc
workclass_ Self-emp-not-inc
workclass_ State-gov
workclass_ Without-pay
education_ 10th
education_ 11th
education_ 12th
education_ 1st-4th
education_ 5th-6th
education_ 7th-8th
education_ 9th
education_ Assoc-acdm
education_ Assoc-voc
education_ Bachelors
education_ Doctorate
education_ HS-grad
education_ Masters
education_ Preschool
education_ Prof-school
education_ Some-college
marital-status_ Divorced
marital-status_ Married-AF-spouse
marital-status_ Married-civ-spouse
marital-status_ Married-spouse-absent
marital-status_ Never-married
marital-status_ Separated
marital-status_ Widowed
occupation_ Adm-clerical
occupation_ Armed-Forces
occupation_ Craft-repair
occupation_ Exec-managerial
occupation_ Farming-fishing
occupation_ Handlers-cleaners
occupation_ Machine-op-inspct
occupation_ Other-service
occupation_ Priv-house-serv
occupat

In [42]:
tmp = X_test.columns
for x in tmp :
    print(x)

age
fnlwgt
education-num
capital-gain
capital-loss
hours-per-week
workclass_ Federal-gov
workclass_ Local-gov
workclass_ Private
workclass_ Self-emp-inc
workclass_ Self-emp-not-inc
workclass_ State-gov
workclass_ Without-pay
education_ 10th
education_ 11th
education_ 12th
education_ 1st-4th
education_ 5th-6th
education_ 7th-8th
education_ 9th
education_ Assoc-acdm
education_ Assoc-voc
education_ Bachelors
education_ Doctorate
education_ HS-grad
education_ Masters
education_ Preschool
education_ Prof-school
education_ Some-college
marital-status_ Divorced
marital-status_ Married-AF-spouse
marital-status_ Married-civ-spouse
marital-status_ Married-spouse-absent
marital-status_ Never-married
marital-status_ Separated
marital-status_ Widowed
occupation_ Adm-clerical
occupation_ Armed-Forces
occupation_ Craft-repair
occupation_ Exec-managerial
occupation_ Farming-fishing
occupation_ Handlers-cleaners
occupation_ Machine-op-inspct
occupation_ Other-service
occupation_ Priv-house-serv
occupat

In [43]:
# 위치 변경을 위해 'native-country_ Holand-Netherlands' 컬럼을 tmp 변수에 넣어줌

tmp = X_train['native-country_ Holand-Netherlands']

In [44]:
X_train.drop('native-country_ Holand-Netherlands', axis = 1, inplace=True)

In [45]:
X_train = pd.concat([X_train, tmp],axis=1)
X_train

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
no                                                                              
1       25  219199              7             0             0              40   
2       39   52978             10             0          1721              55   
3       35  196899             13             0             0              50   
4       64  135527             11             0             0              40   
5       24   60783             10             0             0              70   
...    ...     ...            ...           ...           ...             ...   
29300   23  117789             13             0             0              40   
29301   20  100605              9             0             0              40   
29302   21  372636              9             0             0              40   
29303   18  258474              6             0             0              40   
29304   33  157446              7             0             0              65   

       workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
no                                                                        
1                           0                     0                   1   
2                           0                     0                   1   
3                           0                     0                   1   
4                           0                     0                   1   
5                           0                     0                   1   
...                       ...                   ...                 ...   
29300                       0                     0                   1   
29301                       0                     0                   1   
29302                       0                     0                   1   
29303                       0                     0                   0   
29304                       0                     0                   1   

       workclass_ Self-emp-inc  ...  native-country_ Puerto-Rico  \
no                              ...                                
1                            0  ...                            0   
2                            0  ...                            0   
3                            0  ...                            0   
4                            0  ...                            0   
5                            0  ...                            0   
...                        ...  ...                          ...   
29300                        0  ...                            0   
29301                        0  ...                            1   
29302                        0  ...                            0   
29303                        0  ...                            0   
29304                        0  ...                            0   

       native-country_ Scotland  native-country_ South  \
no                                                       
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
5                             0                      0   
...                         ...                    ...   
29300                         0                      0   
29301                         0                      0   
29302                         0                      0   
29303                         0                      0   
29304                         0                      0   

       native-country_ Taiwan  native-country_ Thailand  \
no                                                        
1                           0                         0   
2                           0                         0   
3                           0                         0   
4                           0             

In [46]:
X_test

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
no                                                                              
29307   29   83003              9             0             0              40   
29308   45   35136             13             0             0              40   
29309   42   64631             13             0             0              40   
29310   41  195821             16             0          1902              40   
29311   17   25690              6             0             0              10   
...    ...     ...            ...           ...           ...             ...   
48838   45  116789              9             0             0              60   
48839   48  185079              9             0             0              50   
48840   63  117473             10          4386             0              40   
48841   18  150817              7             0             0              20   
48842   31  341632             12             0             0              46   

       workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
no                                                                        
29307                       0                     0                   1   
29308                       0                     0                   1   
29309                       0                     0                   0   
29310                       0                     0                   1   
29311                       0                     0                   1   
...                       ...                   ...                 ...   
48838                       0                     0                   0   
48839                       0                     0                   1   
48840                       0                     0                   1   
48841                       0                     0                   1   
48842                       0                     0                   1   

       workclass_ Self-emp-inc  ...  native-country_ Puerto-Rico  \
no                              ...                                
29307                        0  ...                            0   
29308                        0  ...                            0   
29309                        0  ...                            0   
29310                        0  ...                            0   
29311                        0  ...                            0   
...                        ...  ...                          ...   
48838                        0  ...                            0   
48839                        0  ...                            0   
48840                        0  ...                            0   
48841                        0  ...                            0   
48842                        0  ...                            0   

       native-country_ Scotland  native-country_ South  \
no                                                       
29307                         0                      0   
29308                         0                      0   
29309                         0                      0   
29310                         0                      0   
29311                         0                      0   
...                         ...                    ...   
48838                         0                      0   
48839                         0                      0   
48840                         0                      0   
48841                         0                      0   
48842                         0                      0   

       native-country_ Taiwan  native-country_ Thailand  \
no                                                        
29307                       0                         0   
29308                       0                         0   
29309                       0                         0   
29310                       0             

### Scaler 만들기

In [47]:
from sklearn.preprocessing import StandardScaler

In [48]:
scaler = StandardScaler()
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [49]:
transform_X_train = scaler.transform(X_train)
transform_X_train

array([[-1.02303109,  0.27825592, -1.22293678, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [ 0.03370298, -1.29277656, -0.04888118, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [-0.26822104,  0.06748818,  1.12517443, ..., -0.04293546,
        -0.02187796, -0.00606651],
       ...,
       [-1.32495511,  1.72846083, -0.44023305, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [-1.55139812,  0.64946233, -1.61428865, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [-0.41918305, -0.30540059, -1.22293678, ..., -0.04293546,
        -0.02187796, -0.00606651]])

In [50]:
transform_X_test = scaler.transform(X_test)
transform_X_test

array([[-0.72110707, -1.00899623, -0.44023305, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [ 0.48658901, -1.46140965,  1.12517443, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [ 0.26014599, -1.1826386 ,  1.12517443, ..., -0.04293546,
        -0.02187796, -0.00606651],
       ...,
       [ 1.84524709, -0.68320412, -0.04888118, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [-1.55139812, -0.36805437, -1.22293678, ..., -0.04293546,
        -0.02187796, -0.00606651],
       [-0.57014506,  1.43542751,  0.73382256, ..., -0.04293546,
        -0.02187796, -0.00606651]])

### 모델 생성하기

In [51]:
!pip install lightgbm

In [52]:
!pip install xgboost

In [53]:
from sklearn.model_selection import cross_val_score, train_test_split, KFold
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [54]:
from lightgbm import LGBMClassifier 

lgb_model = LGBMClassifier(n_estimators = 1000, learning_rate=0.01, random_state=123)
lgb_model.fit(transform_X_train, y_train)

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.01, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=1000, n_jobs=-1, num_leaves=31, objective=None,
               random_state=123, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

In [55]:
from xgboost import XGBClassifier

xgb_model = XGBClassifier(n_estimators = 100, learning_rate=0.01, random_state = 123)
xgb_model.fit(transform_X_train, y_train)

C:\Users\rnjsx\anaconda3\lib\site-packages\xgboost\sklearn.py:892: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[20:24:59] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.01, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=12, num_parallel_tree=1,
              objective='binary:logistic', random_state=123, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', use_label_encoder=True,
              validate_parameters=1, verbosity=None)

In [56]:
from sklearn.ensemble import RandomForestClassifier

forest_model = RandomForestClassifier(n_estimators=100, random_state= 11)
forest_model.fit(transform_X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=11, verbose=0,
                       warm_start=False)

### 정확도 분석하기

kfold 교차 검증

In [58]:
# ---------------------------------------------

# 정확도

kfold = KFold(n_splits=10, random_state=123, shuffle=True)

acc_result = cross_val_score(lgb_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                        )

print(acc_result.mean())

# -----------------------------------------------

# F1 score

f1_result = cross_val_score(lgb_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result.mean())

# ------------------------------------------

0.8693925338052211
0.7166345576250718


In [60]:
# ---------------------------------------------

# 정확도

kfold = KFold(n_splits=10, random_state=123, shuffle=True)

acc_result = cross_val_score(xgb_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                        )

print(acc_result.mean())

# -----------------------------------------------

# F1 score

f1_result = cross_val_score(xgb_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result.mean())

# ------------------------------------------

C:\Users\rnjsx\anaconda3\lib\site-packages\xgboost\sklearn.py:892: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[20:32:08] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[20:32:12] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[20:32:14] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[20:32:17] WARNING: C:/Users/Administrator/workspace/xgboost-win64_rel

In [61]:
# ---------------------------------------------

# 정확도

kfold = KFold(n_splits=10, random_state=123, shuffle=True)

acc_result = cross_val_score(forest_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                        )

print(acc_result.mean())

# -----------------------------------------------

# F1 score

f1_result = cross_val_score(forest_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result.mean())

# ------------------------------------------

0.8510655527037543
0.6782019919138523
